### ISSUE TITLE
Create SSR Tests

### ISSUE THREAD

**Issue by moog16** [material-components/material-components-web-react#588]
Many issues have been arising due to some components not being compatible with SSR (server side rendering). We should have some tests to avoid this issue. This would also put us up to par with https://github.com/jamesmfriedman/rmwc who has had SSR tests in place.

**Comment by 4cm4k1** [comment]
As I said in #581, I’ll check out RMWC’s testing setup for SSR use cases and report back on it. I don’t have a definite ETA, however. 😄

**Comment by lucasecdb** [comment]
any progress on this? I can't update my packages to versions above v0.7.x because my SSR breaks :-(

let me know if I can help with anything

**Comment by moog16** [comment]
@lucasecdb nothing has been started on this yet. If you'd like to propose something I'm all ears. I currently am not 100% committed to working on this project and am just acting as the maintainer. My plan is to hopefully get back to it in Q2/Q3.

**Comment by lucasecdb** [comment]
@moog16 I have a really bold proposal, to be honest. I see that the majority of the React community and developers like to work more with [writing tests in jest](https://labs.contactually.com/migrating-2-000-specs-from-karma-to-jest-25dd8b0f3cfb) than any other test runner/framework. And it really makes sense when you apply the utilities that come builtin and snapshotting tests that compare what DOM elements the component is "spitting" and *also* run in node, which counts for those SSR tests we want.

The good think about the snapshots is because we could easily compare them with the required DOM structure that the mdc web core module requires, without needing to run any browser and inspecting the elements panel on dev tools.

There are currently a lot of codemods out there which can migrate most of the test suite really quickly, and the ones that are left out can be easily migrated by hand, since there won't be a lot of them. I don't know if you guys already considered rewriting the unit test suite in jest, but I would be willing to work on this migration if you accept this proposal.

**Comment by moog16** [comment]
@lucasecdb I've never used Jest, and I know that the majority of the React community uses it. I understand there are a lot of tools and features that we could use. I am hesitant to switch testing frameworks since it is such a big change. There is also the integration with Travis and the release that would need to be addressed. Can you write out all the different tasks that would be involved? 

Would you be replacing Karma for Jest? I'm not sure how it fits in the picture and haven't looked into the tool. 

TLDR; I'm open to the idea, but need more information.

**Comment by lucasecdb** [comment]
> Would you be replacing Karma for Jest?

Not only Karma, but Chai and Mocha too. Jest already provides a similar toolset that could replace these libraries without any issues. I've seen that you guys also use `testdouble` a lot in the tests, and I do think it'd be harder to replace it at once (because of the lack of codemods), but it's possible to do so using `jest.fn()`, and we could migrate them gradually after the "rewrite".

I've done some research, and seen that the only methods of `assert` that the codemods wouldn't be able to replace are `containsAllKeys`, `deepInclude`, and `sameOrderedMembers`.

> Can you write out all the different tasks that would be involved?

I could open another issue listing them out, what do you think?

**Comment by moog16** [comment]
Couldn't we still use test double with jest? It sounds like we could slowly
migrate that way too. Before you get too deep into the rewrite, let me
consult with the rest of the MDC Web team. There might be some thing I'm
missing.

Otherwise yes a new issue is a good next step.

**Comment by lucasecdb** [comment]
> Couldn't we still use test double with jest?

Yes! The plan would be to slowly migrate it.

> let me consult with the rest of the MDC Web team

Sure! I'll be waiting :-)

**Comment by moog16** [comment]
@lucasecdb I don't think there is any reason we can't try to use jest. If you want to give it a shot I say go for it. But this task sounds like it could be a black hole, so please be transparent and communicate frequently - especially if you get stuck somewhere.

**Comment by moog16** [comment]
@lucasecdb I am curious about the snapshots (https://jestjs.io/docs/en/snapshot-testing)...will this actually detect the issues that you have been facing?

**Comment by moog16** [comment]
@lucasecdb I am curious about the snapshots (https://jestjs.io/docs/en/snapshot-testing)...will this actually detect the issues that you have been facing?

**Comment by lucasecdb** [comment]
well, on a second thought, I think `jsdom` exposes a mocked `window` as a global variable, which wouldn't help much. I still think the jest migration would be nice, though, as it's what react developers are more comfortable using. But, in the end of the day, it's the maintainers that decide what is best for their codebase.

We could setup these just as rmwc have, which would be to assert that `renderToString` from `react-dom/server` doesn't explode, but that could be easily be done in karma/chai/mocha too.

**Comment by moog16** [comment]
My only hesitation is you converting everything/spending a lot of time on it, and then the maintainers (us) not liking the direction. If you don't mind this, or can piecemeal the changes into digestible chunks I say try it out :)
